File: src/cli/agents_md.py
"""Per-workspace AGENTS.md / CLAUDE.md generation and host-skill staging.

* generate_agents_md emits the in-jail briefing files (one per agent —
  Copilot, Gemini, Claude) into AGENTS_DIR/<cname>/.  The files are
  bind-mounted into the jail at boot; their content is the agent's
  primary source of truth for how this jail is set up.
* _prepare_skills + _copy_skill_subdirs stage per-agent skills dirs
  (skills-copilot/, skills-gemini/, skills-claude/) that get bind-mounted
  :ro at /home/agent/.<agent>/skills.  Each one mirrors its host
  counterpart 1:1 (~/.<agent>/skills/) plus the built-in jail-startup
  skill — no cross-agent merging.
"""

import shutil
import tomllib
from pathlib import Path
from typing import Any, Dict, List, Optional

from .paths import AGENTS_DIR


def _workspace_is_yolo_source_tree(workspace: Path) -> bool:
    """True when the workspace being jailed is itself a yolo-jail source
    checkout: ``src/cli/__init__.py`` present AND ``pyproject.toml``
    naming the ``yolo-jail`` project.  An absent, unreadable, or foreign
    pyproject reads as "not a yolo repo"."""
    if not (workspace / "src" / "cli" / "__init__.py").exists():
        return False
    try:
        data = tomllib.loads((workspace / "pyproject.toml").read_text())
    except (OSError, UnicodeDecodeError, tomllib.TOMLDecodeError):
        return False
    project = data.get("project")
    return isinstance(project, dict) and project.get("name") == "yolo-jail"


def generate_agents_md(
    cname: str,
    workspace: Path,
    blocked_tools: List[Dict[str, Any]],
    mount_descriptions: List[str],
    net_mode: str = "bridge",
    runtime: str = "podman",
    forward_host_ports: Optional[List] = None,
    loopholes: Optional[List[tuple]] = None,
    resources: Optional[Dict[str, Any]] = None,
    agents_md_extra: Optional[str] = None,
) -> Path:
    """Generate per-workspace AGENTS.md and CLAUDE.md files and return the directory.

    Produces separate files for Copilot, Gemini, and Claude so that user-level
    ~/.copilot/AGENTS.md, ~/.gemini/AGENTS.md, and ~/.claude/CLAUDE.md content
    can differ between the agents.

    ``agents_md_extra`` is appended verbatim to the generated jail-managed
    content (before host-level user content is prepended) so per-workspace
    or per-user notes — extra MCP server usage hints, project conventions,
    etc. — can ride along with each agent's briefing.
    """
    agents_dir = AGENTS_DIR / cname
    agents_dir.mkdir(parents=True, exist_ok=True)

    if net_mode == "host":
        network_line = "- **Network**: Host networking — the container shares the host network stack. `localhost` / `127.0.0.1` resolves directly to the host. No port mapping needed."
    else:
        network_line = "- **Network**: Bridge mode. Use `host.containers.internal` (resolves to 169.254.1.2) to reach the host."

    # Build forwarded host ports description
    forwarded_ports_lines = []
    if forward_host_ports and net_mode != "host":
        forwarded_ports_lines.append(
            "- **Forwarded Host Ports**: The following host services are available on `localhost` inside this container:"
        )
        for entry in forward_host_ports:
            if isinstance(entry, int):
                forwarded_ports_lines.append(
                    f"  - `localhost:{entry}` → host port {entry}"
                )
            elif isinstance(entry, str) and ":" in entry:
                parts = entry.split(":", 1)
                forwarded_ports_lines.append(
                    f"  - `localhost:{parts[0]}` → host port {parts[1]}"
                )
            elif isinstance(entry, str):
                forwarded_ports_lines.append(
                    f"  - `localhost:{entry}` → host port {entry}"
                )

    resource_line = []
    if resources:
        limits = ", ".join(f"{k}={v}" for k, v in sorted(resources.items()))
        resource_line = [
            f"- **Resource limits** (kernel-enforced): {limits}.  Sub-limit your own processes with `yolo-cglimit` (`--help` for usage)."
        ]

    # Provisioning-failure breadcrumb: only when the last boot actually
    # failed.  The briefing is refreshed on every yolo invocation, so a
    # failure surfaces here on the next attach even though the initial
    # generation ran before provisioning.
    provisioning_failed = []
    try:
        log_text = (workspace / ".yolo" / "startup.log").read_text(errors="replace")
        if "PROVISIONING FAILED" in log_text:
            provisioning_failed = [
                "## ⚠ Provisioning failed",
                "",
                "The last boot's provisioning failed — project tools may be missing.",
                "Read `/workspace/.yolo/startup.log` and self-serve (e.g. run",
                "`mise install` in /workspace, then re-run the step that failed).",
                "",
            ]
    except OSError:
        pass

    lines = [
        "# YOLO Jail Environment",
        "",
        "You are running inside a YOLO Jail — a sandboxed container.",
        "Jail tooling: `yolo --help`; config reference: `yolo config-ref`.",
        "",
        *provisioning_failed,
        "## Environment",
        "",
        f"- **Workspace**: `/workspace` is the host directory `{workspace}`,",
        "  bind-mounted LIVE — the same files, not a copy. Host-side edits are",
        "  instantly visible here and vice versa; there is never a git",
        "  pull/push, fetch, or any sync step between the jail and the host",
        "  for this directory.",
        "- **Home**: `/home/agent` (persistent across sessions)",
        "- **OS**: NixOS-based minimal container (no systemd, no sudo)",
        network_line,
        *forwarded_ports_lines,
        *resource_line,
        "",
        "⚠ rg is recursive by default — never pass grep-style `-r`/`-rn` flags",
        "(in rg, `-r` means `--replace` and silently corrupts match output).",
        "Use `rg -n <pattern> [path]`.",
        "",
    ]

    if loopholes:
        lines.append("## Loopholes — host capabilities wired into this jail")
        lines.append("")
        for name, desc in loopholes:
            first = (desc or "").split(". ")[0].split("\n")[0].strip().rstrip(".")
            lines.append(f"- **{name}**" + (f": {first}" if first else ""))
        lines.append("")
        lines.append("Details: `yolo loopholes list`.")
        lines.append("")

    if blocked_tools:
        lines.append("## Blocked Tools")
        lines.append("")
        lines.append("The following tools are blocked or shimmed in this project:")
        lines.append("")
        for tool in blocked_tools:
            name = tool.get("name", str(tool))
            msg = tool.get("message", "")
            sug = tool.get("suggestion", "")
            entry = f"- `{name}`"
            if msg:
                entry += f": {msg}"
            if sug:
                entry += f" Use `{sug}` instead."
            lines.append(entry)
        lines.append("")

    if mount_descriptions:
        lines.append("## Additional Context Mounts (read-only)")
        lines.append("")
        for m in mount_descriptions:
            host_path, container_path = m.split(":", 1) if ":" in m else (m, m)
            lines.append(f"- `{container_path}` (from host `{host_path}`)")
        lines.append("")

    lines.extend(
        [
            "## Limitations",
            "",
            "- Full internet access, but no host credentials (no ~/.ssh, no ~/.gitconfig): git push/pull and `gh auth login` will not work.",
            "- No sudo/root; context mounts under `/ctx/` are read-only.",
            "",
            "## Packages & Resource Limits",
            "",
            "To request a tool or a container-limit change: edit `/workspace/yolo-jail.jsonc`",
            "(`packages` / `resources`), ALWAYS run `yolo check` after every config edit",
            "(`yolo check --no-build` is fine inside a running jail), then ask the human to",
            "restart the jail. Reference: `yolo config-ref`.",
            "",
            "## Skills",
            "",
            "User-level skills dirs (`~/.<agent>/skills/`) are **read-only** in-jail",
            "(kernel-enforced); workspace-level ones (`/workspace/.<agent>/skills/`) are",
            "writable — develop there, then ask the human to promote to the host.",
            "",
        ]
    )

    if _workspace_is_yolo_source_tree(workspace):
        lines.extend(
            [
                "## Testing Changes to yolo-jail",
                "",
                "The `/workspace` directory is a bind mount of the host's repo, and it also",
                "backs `/opt/yolo-jail` — so nested jails launched from here run your edited",
                "`src/cli` code live.",
                "",
                "When modifying `src/cli/` or `src/entrypoint/`, **always verify with a nested",
                "jail** before telling the human to test on the host. Run `yolo -- bash` from",
                "inside this jail to launch one and confirm your changes work end-to-end.",
                "Container startup errors (mount failures, permission errors, read-only",
                "filesystem conflicts) are only caught by actually running the container —",
                "unit tests alone are not sufficient.",
                "",
                "**Important:** Changes to `src/cli/` take effect on the next `yolo` invocation",
                "on the host (no rebuild needed). Changes to `src/entrypoint/` or `flake.nix`",
                "require `just load && just install` on the host since the entrypoint is baked",
                "into the Nix image.",
                "",
            ]
        )

    jail_content = "\n".join(lines) + "\n"
    if agents_md_extra:
        extra = agents_md_extra.rstrip() + "\n"
        jail_content = jail_content + "\n" + extra

    home = Path.home()
    for agent, dotdir in [("copilot", ".copilot"), ("gemini", ".gemini")]:
        user_agents = home / dotdir / "AGENTS.md"
        if user_agents.exists():
            user_content = user_agents.read_text()
            content = user_content + "\n---\n\n" + jail_content
        else:
            content = jail_content
        _write_briefing(agents_dir / f"AGENTS-{agent}.md", content)

    # Claude reads ~/.claude/CLAUDE.md (not AGENTS.md) at the user-config level.
    user_claude = home / ".claude" / "CLAUDE.md"
    if user_claude.exists():
        claude_content = user_claude.read_text() + "\n---\n\n" + jail_content
    else:
        claude_content = jail_content
    _write_briefing(agents_dir / "CLAUDE.md", claude_content)

    return agents_dir


def _write_briefing(path: Path, content: str) -> None:
    """Truncate-in-place, but never through a hardlink.

    In-place truncation preserves the inode a running jail's file bind
    mount captured, so refreshes propagate (see _refresh_jail_briefings).
    But a hardlink-dedup pass (`yolo prune`) can fuse identical briefing
    files onto ONE inode — truncating through that link would clobber
    every sibling briefing with whichever agent's content is written
    last (observed 2026-07-04 after a host-side `yolo prune --apply`).
    A multi-linked file has already lost its live-refresh identity, so
    breaking the link with a fresh inode is strictly an improvement.
    """
    try:
        if path.lstat().st_nlink > 1:
            path.unlink()
    except OSError:
        pass
    path.write_text(content)


# ---------------------------------------------------------------------------
# Skills merging (host-side, for :ro bind mounts)
# ---------------------------------------------------------------------------

_BUILTIN_JAIL_STARTUP_SKILL = """\
---
name: jail-startup
description: First-run skill for agents entering a YOLO Jail. Reads the handover document left by the outer agent and orients you to the jail environment. Invoke this skill immediately when starting a new session inside a jail.
---

# Jail Startup

You are running inside a **YOLO Jail** — an isolated container environment.
This skill helps you pick up where the previous (outer) agent left off.

## Step 1: Read the Handover Document

The outer agent was REQUIRED to write a handover document before you were
launched. Read it now:

**Primary location:** `.yolo/handover.md` (i.e., `/workspace/.yolo/handover.md`)

If it exists, read it carefully — it contains:
- What the outer agent was working on
- What remains to be done
- Key decisions and rationale
- Files to look at first
- Gotchas and context you need

If the file does NOT exist, tell the human:
> "No handover document found at `.yolo/handover.md`. The outer agent should
> have created one. Can you tell me what I should be working on?"

## Step 2: Orient Yourself

Key facts about your environment:
- **Workspace** is at `/workspace` — this is the SAME directory as on the host (bind-mounted read-write). Changes you make are immediately visible on the host.
- **Internet** is available. You can curl, pip install, npm install, etc.
- **Home** is `/home/agent` — shared across ALL jail workspaces. Auth tokens, tool caches, and configs persist here.
- **Tools**: git, rg, fd, bat, jq, nvim, curl, gh, uv, mise, tmux, and more.
- **Runtimes**: Node.js, Python, Go (managed by mise).
- **Blocked tools**: Some tools may be shimmed (e.g., grep → rg). Check AGENTS.md or run `ls ~/.yolo-shims/` if you hit unexpected blocks. Set `YOLO_BYPASS_SHIMS=1` for scripts that need originals.
- **No pagers**: `PAGER=cat`. Never pipe to `less` or `more`.
- Run `yolo config-ref` for full configuration and environment reference.

## Step 3: Execute

After reading the handover document, proceed with the tasks described in it.
You have full capability — treat this as your primary working environment.
"""


def _prepare_skills(cname: str) -> Path:
    """Prepare per-agent skills directories on the host for :ro bind mounting.

    Each agent's staging dir mirrors its host counterpart 1:1:
      * skills-copilot/ ← ~/.copilot/skills/
      * skills-gemini/  ← ~/.gemini/skills/
      * skills-claude/  ← ~/.claude/skills/

    Plus the built-in ``jail-startup`` skill in every staging dir (it's
    not a host skill — it's our orientation doc).

    Workspace skills (``<workspace>/.{copilot,gemini,claude}/skills/``)
    are NOT collected here — agents already discover them natively from
    the workspace tree, so duplicating them into the user-level mount
    would surface the same skill twice.

    Returns the staging directory containing skills-copilot/,
    skills-gemini/, skills-claude/.
    """
    staging = AGENTS_DIR / cname
    staging.mkdir(parents=True, exist_ok=True)

    home = Path.home()

    for agent_suffix in ("copilot", "gemini", "claude"):
        skills_dir = staging / f"skills-{agent_suffix}"
        skills_dir.mkdir(exist_ok=True)
        # Clear contents *inside* skills_dir — never unlink skills_dir
        # itself.  `run_cmd` bind-mounts this exact path into the running
        # container at /home/agent/.<agent>/skills.  Linux bind mounts
        # capture the source's *inode* at mount time, not its path; if
        # we rmtree(skills_dir)+mkdir(skills_dir) here, mkdir allocates
        # a new inode and the container's mount stays pinned to the old
        # one — so attach-time refreshes (the whole point of the
        # `_refresh_jail_briefings` plumbing) silently no-op for any
        # running jail.  Removing entries *inside* the dir is fine: the
        # parent mount is unaffected and the container sees the updated
        # listing immediately.
        for child in skills_dir.iterdir():
            if child.is_dir():
                shutil.rmtree(child)
            else:
                child.unlink()

        # 1. Built-in skills (every agent gets jail-startup).
        builtin = skills_dir / "jail-startup"
        builtin.mkdir()
        (builtin / "SKILL.md").write_text(_BUILTIN_JAIL_STARTUP_SKILL)

        # 2. Host user-level skills — strictly per-agent.  Whatever's in
        # ~/.<agent>/skills/ is what the matching jail agent sees; no
        # cross-agent merging.  Deleting from one host dir cleanly
        # removes from the matching jail dir.
        _copy_skill_subdirs(home / f".{agent_suffix}" / "skills", skills_dir)

    return staging


def _copy_skill_subdirs(src: Path, dst: Path):
    """Copy skill subdirectories from src into dst, following symlinks."""
    if not src.is_dir():
        return
    for item in src.iterdir():
        if item.is_dir():
            target = dst / item.name
            if target.exists():
                shutil.rmtree(target)
            shutil.copytree(item, target, symlinks=False)

File: internal/agents/agentsmd.go
package agents

// Per-workspace AGENTS.md / CLAUDE.md briefing generation and host-skill
// staging. The briefing content is a byte-exact string contract;
// WriteBriefing's hardlink-breaking truncation is an inode-preservation
// contract a running jail's bind mount depends on.

import (
	"bytes"
	"os"
	"sort"
	"strings"
)

// BlockedTool is one entry of the "Blocked Tools" section (name + optional
// message + optional suggestion).
type BlockedTool struct {
	Name       string
	Message    string
	Suggestion string
}

// Loophole is a (name, description) pair for the loopholes section.
type Loophole struct {
	Name string
	Desc string
}

// BriefingInput carries everything generate_agents_md's jail-managed content
// depends on. Workspace is the host workspace path (rendered verbatim);
// ProvisioningFailed is true when the last boot's .yolo/startup.log contained
// "PROVISIONING FAILED" (the caller reads the log — see ReadProvisioningFailed).
type BriefingInput struct {
	Workspace          string
	BlockedTools       []BlockedTool
	MountDescriptions  []string
	NetMode            string
	ForwardHostPorts   []any
	Loopholes          []Loophole
	Resources          map[string]any
	IsYoloSourceTree   bool
	ProvisioningFailed bool
}

// BriefingContent renders the jail-managed briefing body (before any host-level
// user content is prepended and before agents_md_extra is appended). This is
// the byte-exact string generate_agents_md builds via its `lines` list joined
// with "\n" plus a trailing newline. NetMode defaults to "bridge" when empty.
func BriefingContent(in BriefingInput) string {
	netMode := in.NetMode
	if netMode == "" {
		netMode = "bridge"
	}

	var networkLine string
	if netMode == "host" {
		networkLine = "- **Network**: Host networking — the container shares the host network stack. `localhost` / `127.0.0.1` resolves directly to the host. No port mapping needed."
	} else {
		networkLine = "- **Network**: Bridge mode. Use `host.containers.internal` (resolves to 169.254.1.2) to reach the host."
	}

	var forwardedPorts []string
	if len(in.ForwardHostPorts) > 0 && netMode != "host" {
		forwardedPorts = append(forwardedPorts,
			"- **Forwarded Host Ports**: The following host services are available on `localhost` inside this container:")
		for _, entry := range in.ForwardHostPorts {
			lp, hp, kind := portEntry(entry)
			switch kind {
			case portInt:
				forwardedPorts = append(forwardedPorts, "  - `localhost:"+lp+"` → host port "+lp)
			case portMapped:
				forwardedPorts = append(forwardedPorts, "  - `localhost:"+lp+"` → host port "+hp)
			case portPlain:
				forwardedPorts = append(forwardedPorts, "  - `localhost:"+lp+"` → host port "+lp)
			}
		}
	}

	var resourceLine []string
	if len(in.Resources) > 0 {
		keys := make([]string, 0, len(in.Resources))
		for k := range in.Resources {
			keys = append(keys, k)
		}
		sort.Strings(keys)
		parts := make([]string, 0, len(keys))
		for _, k := range keys {
			parts = append(parts, k+"="+pyValue(in.Resources[k]))
		}
		resourceLine = []string{
			"- **Resource limits** (kernel-enforced): " + strings.Join(parts, ", ") +
				".  Sub-limit your own processes with `yolo-cglimit` (`--help` for usage).",
		}
	}

	var provisioningFailed []string
	if in.ProvisioningFailed {
		provisioningFailed = []string{
			"## ⚠ Provisioning failed",
			"",
			"The last boot's provisioning failed — project tools may be missing.",
			"Read `/workspace/.yolo/startup.log` and self-serve (e.g. run",
			"`mise install` in /workspace, then re-run the step that failed).",
			"",
		}
	}

	lines := []string{
		"# YOLO Jail Environment",
		"",
		"You are running inside a YOLO Jail — a sandboxed container.",
		"Jail tooling: `yolo --help`; config reference: `yolo config-ref`.",
		"",
	}
	lines = append(lines, provisioningFailed...)
	lines = append(lines,
		"## Environment",
		"",
		"- **Workspace**: `/workspace` is the host directory `"+in.Workspace+"`,",
		"  bind-mounted LIVE — the same files, not a copy. Host-side edits are",
		"  instantly visible here and vice versa; there is never a git",
		"  pull/push, fetch, or any sync step between the jail and the host",
		"  for this directory.",
		"- **Home**: `/home/agent` (persistent across sessions)",
		"- **OS**: NixOS-based minimal container (no systemd, no sudo)",
		networkLine,
	)
	lines = append(lines, forwardedPorts...)
	lines = append(lines, resourceLine...)
	lines = append(lines,
		"",
		"⚠ rg is recursive by default — never pass grep-style `-r`/`-rn` flags",
		"(in rg, `-r` means `--replace` and silently corrupts match output).",
		"Use `rg -n <pattern> [path]`.",
		"",
	)

	if len(in.Loopholes) > 0 {
		lines = append(lines, "## Loopholes — host capabilities wired into this jail", "")
		for _, lh := range in.Loopholes {
			first := loopholeFirst(lh.Desc)
			if first != "" {
				lines = append(lines, "- **"+lh.Name+"**: "+first)
			} else {
				lines = append(lines, "- **"+lh.Name+"**")
			}
		}
		lines = append(lines, "", "Details: `yolo loopholes list`.", "")
	}

	if len(in.BlockedTools) > 0 {
		lines = append(lines,
			"## Blocked Tools",
			"",
			"The following tools are blocked or shimmed in this project:",
			"",
		)
		for _, tool := range in.BlockedTools {
			entry := "- `" + tool.Name + "`"
			if tool.Message != "" {
				entry += ": " + tool.Message
			}
			if tool.Suggestion != "" {
				entry += " Use `" + tool.Suggestion + "` instead."
			}
			lines = append(lines, entry)
		}
		lines = append(lines, "")
	}

	if len(in.MountDescriptions) > 0 {
		lines = append(lines, "## Additional Context Mounts (read-only)", "")
		for _, m := range in.MountDescriptions {
			hostPath, containerPath := m, m
			if i := strings.Index(m, ":"); i >= 0 {
				hostPath, containerPath = m[:i], m[i+1:]
			}
			lines = append(lines, "- `"+containerPath+"` (from host `"+hostPath+"`)")
		}
		lines = append(lines, "")
	}

	lines = append(lines,
		"## Limitations",
		"",
		"- Host credentials are not propagated into the jail: the host's `~/.ssh`,",
		"  `~/.gitconfig`, and cloud/gh tokens are invisible here. This is a credential",
		"  boundary, not a network block — outbound SSH and HTTPS work normally, so git",
		"  push/pull and API calls succeed whenever the jail has its own credentials",
		"  (e.g. a workspace-specific deploy key or a token in `.env`). Only without",
		"  such jail-local credentials do authenticated operations fail.",
		"- No sudo/root; context mounts under `/ctx/` are read-only.",
		"",
		"## Packages & Resource Limits",
		"",
		"To request a tool or a container-limit change: edit `/workspace/yolo-jail.jsonc`",
		"(`packages` / `resources`), ALWAYS run `yolo check` after every config edit",
		"(`yolo check --no-build` is fine inside a running jail), then ask the human to",
		"restart the jail. Reference: `yolo config-ref`.",
		"",
		"## Skills",
		"",
		"User-level skills dirs (`~/.<agent>/skills/`) are **read-only** in-jail",
		"(kernel-enforced); workspace-level ones (`/workspace/.<agent>/skills/`) are",
		"writable — develop there, then ask the human to promote to the host.",
		"",
	)

	if in.IsYoloSourceTree {
		lines = append(lines,
			"## Testing Changes to yolo-jail",
			"",
			"The `/workspace` directory is a bind mount of the host's repo, and it also",
			"backs `/opt/yolo-jail` — so nested jails launched from here run your edited",
			"Go code live (via dev-override wrappers that prefer `/opt/yolo-jail/dist-go/`).",
			"",
			"When modifying `cmd/` or `internal/`, **always verify with a nested",
			"jail** before telling the human to test on the host. Run `yolo -- bash` from",
			"inside this jail to launch one and confirm your changes work end-to-end.",
			"Container startup errors (mount failures, permission errors, read-only",
			"filesystem conflicts) are only caught by actually running the container —",
			"unit tests alone are not sufficient.",
			"",
			"**Important:** Run `just deploy` to cross-compile Go binaries. Changes to",
			"`flake.nix` require `just load && just install` on the host since the image",
			"is baked by Nix.",
			"",
		)
	}

	return strings.Join(lines, "\n") + "\n"
}

// ComposeBriefing appends agents_md_extra to the jail content the way
// generate_agents_md does: jail_content + "\n" + extra.rstrip() + "\n" when
// extra is non-empty.
func ComposeBriefing(jailContent, extra string) string {
	if extra == "" {
		return jailContent
	}
	return jailContent + "\n" + strings.TrimRight(extra, " \t\r\n") + "\n"
}

// PrependHostBriefing produces one agent's final briefing: the host briefing
// file's content + "\n---\n\n" + jailContent when the host file exists, else
// jailContent alone.
func PrependHostBriefing(hostBriefingPath, jailContent string) string {
	data, err := os.ReadFile(hostBriefingPath)
	if err != nil {
		return jailContent
	}
	return string(data) + "\n---\n\n" + jailContent
}

type portKind int

const (
	portNone portKind = iota
	portInt
	portMapped
	portPlain
)

// portEntry classifies a forward_host_ports entry the way generate_agents_md's
// isinstance ladder does, returning the rendered local/host port strings. An
// int → (n, n, portInt); a string "a:b" → (a, b, portMapped) [split once]; a
// plain string → (s, s, portPlain); anything else → portNone.
func portEntry(entry any) (local, host string, kind portKind) {
	// jsonx decodes ints as jsonInt; accept both that and native ints.
	if s, ok := intString(entry); ok {
		return s, s, portInt
	}
	if str, ok := entry.(string); ok {
		if i := strings.Index(str, ":"); i >= 0 {
			return str[:i], str[i+1:], portMapped
		}
		return str, str, portPlain
	}
	return "", "", portNone
}

// loopholeFirst extracts the first-sentence summary of a loophole description,
// mirroring: (desc or "").split(". ")[0].split("\n")[0].strip().rstrip(".").
func loopholeFirst(desc string) string {
	s := desc
	if i := strings.Index(s, ". "); i >= 0 {
		s = s[:i]
	}
	if i := strings.Index(s, "\n"); i >= 0 {
		s = s[:i]
	}
	s = strings.TrimSpace(s)
	return strings.TrimRight(s, ".")
}

// WorkspaceIsYoloSourceTree reports whether workspace is a yolo-jail source
// checkout: go.mod with the yolo-jail module path AND cmd/yolo/main.go present.
func WorkspaceIsYoloSourceTree(workspace string) bool {
	data, err := os.ReadFile(workspace + "/go.mod")
	if err != nil {
		return false
	}
	if !bytes.Contains(data, []byte("yolo-jail")) {
		return false
	}
	_, err = os.Stat(workspace + "/cmd/yolo/main.go")
	return err == nil
}
